packages/components/src/range-control/utils.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+import { clamp, noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useRef, useEffect, useState } from '@wordpress/element';
+
+/**
+ * A float supported clamp function for a specific value.
+ *
+ * @param {number} value The value to clamp
+ * @param {number} min The minimum value
+ * @param {number} max The maxinum value
+ * @return {number} A (float) number
+ */
+function floatClamp( value, min, max ) {
+	return parseFloat( clamp( value, min, max ) );
+}
+
+/**
+ * Hook to store a clamped value, derived from props.
+ */
+export function useControlledRangeValue( { min, max, value: valueProp = 0 } ) {
+	const [ value, _setValue ] = useState( floatClamp( valueProp, min, max ) );
+	const valueRef = useRef( value );
+
+	const setValue = useCallback(
+		( nextValue ) => {
+			_setValue( floatClamp( nextValue, min, max ) );
+		},
+		[ _setValue, min, max ]
+	);
+
+	useEffect( () => {
+		if ( valueRef.current !== valueProp ) {
+			setValue( valueProp );
+			valueRef.current = valueProp;
+		}
+	}, [ valueRef, valueProp, setValue ] );
```

Comment:
It's a bit unclear to me, but I _think_ you generally won't want to include a (local) ref in the dependencies array.

Related:

- https://github.com/facebook/react/issues/16121#issuecomment-511369830
- https://github.com/facebook/react/issues/14387#issuecomment-503616820

Reply:
Interesting!! I've been unclear about this as well, as I've seen both implementations (not specifically in Gutenberg, just in general). Thank you for sending those links!~ <3

Will refactor to remove local ref